streams/test-utils/src/test/java/org/apache/kafka/streams/test/TopologyTestDriverTest.java

```diff
@@ -0,0 +1,573 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.test;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.SystemTime;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.Stores;
+import org.apache.kafka.streams.state.internals.KeyValueStoreBuilder;
+import org.apache.kafka.test.TestUtils;
+import org.junit.After;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class TopologyTestDriverTest {
+    private final static String SOURCE_TOPIC_1 = "source-topic-1";
+    private final static String SOURCE_TOPIC_2 = "source-topic-2";
+    private final static String SINK_TOPIC_1 = "sink-topic-1";
+    private final static String SINK_TOPIC_2 = "sink-topic-2";
+
+    private final ConsumerRecordFactory<byte[], byte[]> consumerRecordFactory = new ConsumerRecordFactory<>(
+        new ByteArraySerializer(),
+        new ByteArraySerializer());
+
+    private final byte[] key1 = new byte[0];
+    private final byte[] value1 = new byte[0];
+    private final long timestamp1 = 42L;
+    private final ConsumerRecord<byte[], byte[]> consumerRecord1 = consumerRecordFactory.create(SOURCE_TOPIC_1, key1, value1, timestamp1);
+
+    private final byte[] key2 = new byte[0];
+    private final byte[] value2 = new byte[0];
+    private final long timestamp2 = 43L;
+    private final ConsumerRecord<byte[], byte[]> consumerRecord2 = consumerRecordFactory.create(SOURCE_TOPIC_2, key2, value2, timestamp2);
+
+    private TopologyTestDriver testDriver;
+    private final Properties config = new Properties() {
+        {
+            put(StreamsConfig.APPLICATION_ID_CONFIG, "test-TopologyTestDriver");
+            put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234");
+            put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
+        }
+    };
+
+    private final static class Record {
```

Comment:
Why not just using the ConsumerRecord directly?

Reply:
`ConsumerRecord`  does not implement `equals` or `hashCode` and we thus cannot use `assertThat` if we use `ConsumerRecord`.

We are also just interested in some fields, for example we don't need `partition`, `headers`, `serializedKeySize`, etc.